File: play_text_demo/cpp/play_text_demo/src/play_text_node.cpp
/**
 * @file play_text_node.cpp
 * @brief 通过 /intelligent_interaction/tts/play 服务实现文字转语音播放
 *
 * 支持三种资源类型：文本(text)、本地音频文件(file)、远端URL(url)
 */

#include <chrono>
#include <memory>
#include <string>

#include "interaction_msgs/srv/tts_service.hpp"
#include "rclcpp/rclcpp.hpp"

class PlayTextNode : public rclcpp::Node {
 public:
  PlayTextNode() : Node("play_text_demo") {
    this->declare_parameter("text", "你好，我是天工机器人，很高兴认识你。");
    this->declare_parameter("type", "text");   // text / file / url
    this->declare_parameter("cmd", "append");  // append / stop / query

    client_ = this->create_client<interaction_msgs::srv::TtsService>(
        "/intelligent_interaction/tts/play");

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(500),
        std::bind(&PlayTextNode::call_tts_service, this));

    RCLCPP_INFO(this->get_logger(), "PlayTextNode initialized");
    RCLCPP_INFO(this->get_logger(), "Text: %s",
                this->get_parameter("text").as_string().c_str());
  }

 private:
  void call_tts_service() {
    timer_->cancel();

    while (!client_->wait_for_service(std::chrono::milliseconds(1000))) {
      if (!rclcpp::ok()) {
        RCLCPP_ERROR(this->get_logger(),
                     "Interrupted while waiting for TTS service. Exiting.");
        return;
      }
      RCLCPP_INFO(this->get_logger(), "TTS service not available, waiting...");
    }

    auto request = std::make_shared<interaction_msgs::srv::TtsService::Request>();
    request->text = this->get_parameter("text").as_string();
    request->type = this->get_parameter("type").as_string();
    request->cmd = this->get_parameter("cmd").as_string();

    RCLCPP_INFO(this->get_logger(),
                "Calling TTS service: text=\"%s\", type=%s, cmd=%s",
                request->text.c_str(), request->type.c_str(),
                request->cmd.c_str());

    auto future_result = client_->async_send_request(
        request, std::bind(&PlayTextNode::response_callback, this,
                           std::placeholders::_1));
  }

  void response_callback(
      rclcpp::Client<interaction_msgs::srv::TtsService>::SharedFuture future) {
    auto response = future.get();
    if (response->success) {
      RCLCPP_INFO(this->get_logger(), "TTS service success. Status: %s",
                  response->status.c_str());
    } else {
      RCLCPP_ERROR(this->get_logger(), "TTS service failed. Status: %s",
                   response->status.c_str());
    }
  }

  rclcpp::Client<interaction_msgs::srv::TtsService>::SharedPtr client_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PlayTextNode>());
  rclcpp::shutdown();
  return 0;
}

File: 3.play_text_demo/python/play_text_demo/play_text_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Play Text Node (Python version)

通过调用 /intelligent_interaction/tts/play 服务实现文字转语音播放。
支持三种资源类型：文本(text)、本地音频文件(file)、远端URL(url)。
"""

import rclpy
from rclpy.node import Node
from interaction_msgs.srv import TtsService


class PlayTextNode(Node):
    """ROS2 Node for TTS text playback via interaction_msgs/TtsService"""

    def __init__(self):
        super().__init__('play_text_demo')

        self.declare_parameter('text', '你好，我是天工机器人，很高兴认识你。')
        self.declare_parameter('type', 'text')  # text / file / url
        self.declare_parameter('cmd', 'append')  # append / stop / query

        self.client = self.create_client(
            TtsService, '/intelligent_interaction/tts/play')

        self.timer = self.create_timer(0.5, self.call_tts_service)

        self.get_logger().info('PlayTextNode initialized')
        self.get_logger().info(
            f'Text: {self.get_parameter("text").get_parameter_value().string_value}')

    def call_tts_service(self):
        self.timer.cancel()

        while not self.client.wait_for_service(timeout_sec=1.0):
            if not rclpy.ok():
                self.get_logger().error('Interrupted while waiting for TTS service. Exiting.')
                return
            self.get_logger().info('TTS service not available, waiting...')

        request = TtsService.Request()
        request.text = self.get_parameter('text').get_parameter_value().string_value
        request.type = self.get_parameter('type').get_parameter_value().string_value
        request.cmd = self.get_parameter('cmd').get_parameter_value().string_value

        self.get_logger().info(
            f'Calling TTS service: text="{request.text}", type={request.type}, cmd={request.cmd}')

        self.future = self.client.call_async(request)
        self.future.add_done_callback(self.response_callback)

    def response_callback(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(
                    f'TTS service success. Status: {response.status}')
            else:
                self.get_logger().error(
                    f'TTS service failed. Status: {response.status}')
        except Exception as e:
            self.get_logger().error(f'Service call failed: {e}')


def main(args=None):
    rclpy.init(args=args)
    node = PlayTextNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
